### PR TITLE
[FLINK-30257] fix SqlClientITCase#testMatchRecognize

### PR DESCRIPTION
**Pull request by WencongLiu**
## What is the purpose of the change

*fix SqlClientITCase#testMatchRecognize*

## Brief change log

  - *make verifyNumberOfResultRecords wait until getting the expected number of records*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-end-to-end-tests/flink-sql-client-test/src/test/java/SqlClientITCase.java [hunk 1]
```diff
@@ -290,10 +290,15 @@ private void verifyNumberOfResultRecords(String resultFilePath, int expectedNumb
         File tempOutputFile = new File(tempDir, "records.out");
         String tempOutputFilepath = tempOutputFile.toString();
         GenericContainer<?> taskManager = flink.getTaskManagers().get(0);
-        Thread.sleep(5000); // prevent NotFoundException: Status 404
-        taskManager.copyFileFromContainer(resultFilePath, tempOutputFilepath);
-
-        int numberOfResultRecords = UpsertTestFileUtil.getNumberOfRecords(tempOutputFile);
+        int numberOfResultRecords;
+        while (true) {
+            Thread.sleep(5000); // prevent NotFoundException: Status 404
```
Comment (MartijnVisser) [inline flink-end-to-end-tests/flink-sql-client-test/src/test/java/SqlClientITCase.java:295]: @zentol Any thoughts if/how we could get rid of the wait?

Comment (WencongLiu) [reply flink-end-to-end-tests/flink-sql-client-test/src/test/java/SqlClientITCase.java:295]: cc @zentol

Comment (zentol) [reply flink-end-to-end-tests/flink-sql-client-test/src/test/java/SqlClientITCase.java:295]: At the end of the way you will need to poll _something_, be it for the output file or the status of the job.

But the sleep could be way lower; 50ms would be perfectly fine.

Comment (MartijnVisser) [reply flink-end-to-end-tests/flink-sql-client-test/src/test/java/SqlClientITCase.java:295]: @WencongLiu Could you lower the sleep?

Comment (WencongLiu) [reply flink-end-to-end-tests/flink-sql-client-test/src/test/java/SqlClientITCase.java]: Done. thx @MartijnVisser @zentol

**Comment by MartijnVisser** [comment]
@WencongLiu You'll need to rebase your PR with the latest changes from `master` to get the CI to pass.

**Comment by WencongLiu** [comment]
@flinkbot rerun

**Comment by WencongLiu** [comment]
@flinkbot re-run

**Comment by WencongLiu** [comment]
@flinkbot run azure

**Comment by WencongLiu** [comment]
@MartijnVisser Done.

**Review by MartijnVisser — APPROVED** [review]
@WencongLiu Thanks! Will you also create backports to `release-1.17` and potentially `release-1.16` ?

**Comment by WencongLiu** [comment]
> @WencongLiu Thanks! Will you also create backports to `release-1.17` and potentially `release-1.16` ?

Do you mean to create the same pull request for `release-1.17` and `release-1.16`?

**Pull request merged**